### PR TITLE
Fill `ByteBuffer` from `Fd`s too

### PR DESCRIPTION
**Pull request by bitonic**
Note: this is not good for merge yet because things are still not in working order on Windows. We can review the code in the meantime.

Edit: I've given up on Windows, see https://github.com/fpco/store/pull/82/commits/250d9c544b087e03da8ac6d50fca811c99efd4f6

### CODE REVIEW

**Comment by bitonic** [comment]
@mgsloan : I'm not too clear on how the hpack thing works -- do I have to update `package.yaml` too?

File: src/Data/Store/Streaming.hs [hunk 1]
```diff
@@ -86,91 +98,99 @@ encodeMessage (Message x) =
...
-
--- | Decode a 'Message' of known size from a 'ByteBuffer'.
-peekSized :: (MonadIO m, Store a) => ByteBuffer -> Int -> m (PeekMessage m a)
-peekSized bb n =
-    BB.unsafeConsume bb n >>= \case
-        Right ptr -> liftM (Done . Message) $ decodeFromPtr ptr n
-        Left _ -> return $ NeedMoreInput (\ bs -> BB.copyByteString bb bs
-                                                  >> peekSized bb n)
+-- successfully deserialised object, or a request for more input.
+type PeekMessage i m a = FT ((->) i) m a
```
Comment (bitonic) [inline src/Data/Store/Streaming.hs]: Note: `FT`'s API is like that of `FreeT`, but `FT` is faster -- see https://www.stackage.org/haddock/lts-7.3/free-4.12.4/Control-Monad-Trans-Free-Church.html

File: src/Data/Store/Streaming.hs [hunk 2]
```diff
@@ -86,91 +98,99 @@ encodeMessage (Message x) =
...
-            Just bs -> BB.copyByteString bb bs >> decodeSized bb getBs n
-            Nothing -> BB.availableBytes bb >>= \ available ->
-                liftIO $ tooManyBytes n available "Data.Store.Message.Message"
-{-# INLINE decodeSized #-}
+decodeMessage :: (Store a, MonadIO m) => FillByteBuffer i m -> ByteBuffer -> m (Maybe i) -> m (Maybe (Message a))
+decodeMessage fill bb getInp = runMaybeT $
+  iterTM (\consumeInp -> consumeInp =<< MaybeT getInp) (peekMessage fill bb) <|>
+  (do available <- BB.availableBytes bb
+      unless (available == 0) $
+        liftIO $ tooManyBytes available available "Data.Store.Message.Message"
```
Comment (bitonic) [inline src/Data/Store/Streaming.hs]: `liftIO $ tooManyBytes available available "Data.Store.Message.Message"` is wrong, the first argument to `tooManyBytes` should be the expected ones. I'll fix it soon.

File: src/Data/Store/Streaming.hs [hunk 3]
```diff
@@ -1,6 +1,8 @@
 {-# LANGUAGE LambdaCase #-}
 {-# LANGUAGE FlexibleContexts #-}
 {-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE BangPatterns #-}
```
Comment (kantp) [inline src/Data/Store/Streaming.hs]: I think the `BangPatterns` extension is not actually used.

File: src/Data/Store/Streaming.hs [hunk 4]
```diff
@@ -86,91 +101,103 @@ encodeMessage (Message x) =
...
-
--- | Decode a 'Message' of known size from a 'ByteBuffer'.
-peekSized :: (MonadIO m, Store a) => ByteBuffer -> Int -> m (PeekMessage m a)
-peekSized bb n =
-    BB.unsafeConsume bb n >>= \case
-        Right ptr -> liftM (Done . Message) $ decodeFromPtr ptr n
-        Left _ -> return $ NeedMoreInput (\ bs -> BB.copyByteString bb bs
-                                                  >> peekSized bb n)
+-- successfully deserialised object, or a request for more input.
+type PeekMessage i m a = FT ((->) i) m a
```
Comment (kantp) [inline src/Data/Store/Streaming.hs]: It wouldn't hurt to add a comment here, explaining how `FT` is what we want. The new `PeekMessage` is much more elegant and versatile, but a bit less obvious.

**Comment by mgsloan** [comment]
@Bitonic yes, only update the package.yaml

**Comment by mgsloan** [comment]
LGTM!  Good stuff!

**Comment by bitonic** [comment]
Note: this merge didn't really do anything, because it was against `windows-test`. I'll reopen another PR.

**Comment by mgsloan** [comment]
Ah, I didn't notice that, good catch

**Pull request merged**